### DISCUSSION TITLE
Use go-orav2 on Oracle Autonomous Database

### DISCUSSION THREAD

**Discussion by lucasjellema** [sijms/go-ora#145] [Q&A] [answered]
Thanks for the work you put into go-ora. It is really impressive. I appreciate the effort! 

I have a question on using go-orav2 for connecting to an Oracle Cloud Autonomous Database. The short and simple question is: do you have a code example that has been proven to work with Autonomous Database on Oracle Cloud? Or even: is it known if go-ora can be used with OCI Autonomous Database?

The longer story: After successfully using go-orav2 for working with a local database (XE 21c), I now want to try to connect to an ATP instance (Autonomous Transaction Processing). The local connection was based simply on username, password, host, port and servicename. To connect to ATP, I bring in an Oracle Wallet. I have used that same Oracle Wallet file to successfully connect to the ATP instance from the Oracle Developer extension in VS Code - I know the wallet itself is ok.

My code is shown below. Using v2@v2.2.17 it fails on conn.Open(); the error is EOF and it originates in line 476 (call to session.conn.Read(head)) in session.go in package network)

`package main

import (
	"database/sql/driver"
	"fmt"
	go_ora "github.com/sijms/go-ora/v2"
)

func main() {
	//	service := "daprdb_low"
	service := "k8j2fvxbaujdcfy_daprdb_low.adb.oraclecloud.com"
	user := "demo"
	server := "adb.us-ashburn-1.oraclecloud.com"
	port := "1522"
	pw := "Modem123mode"
	wallet := "/home/lucas/dapr-work/components-contrib/state/oracledatabase/Wallet_daprDB"
	connStr := "oracle://" + user + ":" + pw + "@" + server + ":" + port + "/" + service
	connStr += "?wallet=" + wallet
	conn, err := go_ora.NewConnection(connStr)
	if err != nil {
		fmt.Println("Can't create connection: ", err)
		return
	}
	err = conn.Open()
	if err != nil {
		fmt.Println("Can't open connection: ", err)
		return
	}
	defer func() {
		err = conn.Close()
		if err != nil {
			fmt.Println("Can't close connection: ", err)
		}
	}()
}
` 

Do you have any idea what might be the issue? What I can try to do to investigate and make some progress? 

Thanks for any suggestions.
Note: once I have it resolved, I intend to write a blog article and submit a PR with an example to this go-ora repo.

**Comment by sijms** [comment, marked as answer]
thank you for your appreciation
usually oracle cloud using TCPS for connection thus you need to do the following

```Golang
urlOptions := map[string]string{
   //"TRACE FILE": "trace.log",
   "SSL":        "enable",
   "SSL Verify": "false",
   "WALLET": "your wallet path",
}
databaseURL := go_ora.BuildUrl("host",  port,  "service",  "user", "password", urlOptions)
```

**Reply by lucasjellema** [reply]
Dear @sijms ,

Thanks for your rapid response. I have tried - I believe - to follow your guidance.  Unfortunately, the result is the same: Can't open connection:  EOF

Here is my code:
`package main
import (
	"fmt"
	go_ora "github.com/sijms/go-ora/v2"
)

func main() {
	service := "k8j2fvxbaujdcfy_daprdb_low.adb.oraclecloud.com"
	user := "demo"
	server := "adb.us-ashburn-1.oraclecloud.com"
	pw := "Modem123mode"
	wallet := "/home/lucas/dapr-work/components-contrib/state/oracledatabase/Wallet_daprDB/"
	urlOptions := map[string]string{
		//"TRACE FILE": "trace.log",
		"SSL":        "enable",
		"SSL Verify": "false",
		"WALLET":     wallet,
	}
	databaseURL := go_ora.BuildUrl(server, 1522, service, user, pw, urlOptions)
	go_ora.NewConnection(databaseURL)
	conn, err := go_ora.NewConnection(connStr)
	if err != nil {
		fmt.Println("Can't create connection: ", err)
		return
	}
	err = conn.Open()
	if err != nil {
		fmt.Println("Can't open connection: ", err)
		return
	}
`
Do you any obvious mistake in my code? Is there any investigation/analysis you can recommend?
thanks!
Lucas

**Comment by sijms** [comment]
the code is correct
can you uncoment the TRACE FILE line to see the trace
also please note that when you are using wallet for credentials you don't need to supply password as password is stored in the wallet
thus password will be empty

**Reply by lucasjellema** [reply]
A trace file gets created - but it is empty unfortunately.

When I set the password to an empty string, I get the following result: "_Can't create connection:  cannot find credentials for server: , service: k8j2fvxbaujdcfy_daprdb_low.adb.oraclecloud.com,  username: admin_"
I am surprised at the message and also at the first part: server: , - almost as if the server name is not processed correctly.

By the way: what if I want to connect to a different user than the one for which the wallet contains the password? Or is that a scenario not supported with a wallet? Should all users one wants to connect to be defined in the wallet?

Just to be sure: using the GoDror package and the Oracle Instant Client (not my preferred way of connecting obviously) I was able to connect with this wallet, from my laptop to the Autonomous Database.

**Reply by sijms** [reply]
this message means that wallet doesn't contain credential for user admin thus you need to supply password for this user.

the strange thing why trace file is empty

**Comment by sijms** [comment]
when I try your code I get the following error:
`remote error: tls: handshake failure`
this means that SSL certificate that stored in the wallet is required

**Reply by lucasjellema** [reply]
I am not sure what that means for me. I believe there is an SSL certificate stored in the wallet. Do I have to make that explicitly available in my code somehow? Or should the go_ora code be able to retrieve the certificate from the wallet?  What is your suggestion I now should try? 

 BTW: I have found [the line ](https://github.com/sijms/go-ora/blob/v2.2.17/v2/connection_string.go#L316)that produces the message I had before about the missing  credentials

**Reply by sijms** [reply]
no wallet is analyzed and credentials and ssl certificates is read from it automatically once you pass databaseURL to create connection

**Reply by lucasjellema** [reply]
If understand you correctly: the credentials and SSL certificates are read from the wallet automatically. There is no way (or need) for me to trigger that action. And if the wallet is known to be correct (because it works with GoDrOr), this action should be successful. I am at a loss as to what I can do to further investigate the issue. Perhaps create a clean environment from scratch and try all steps again?

Would you happen to have a working example of connecting based on a wallet?
Would you be interested in me sending my wallet file and analyzing this issue further? (the database in question is a free tier ATP instance only set up for experiments and without any real data).

Once again, thanks for your time and attempts to help me out.

**Reply by sijms** [reply]
yes wallet is analyzed automatically
and i will be happy if you send wallet i will try to investigate the issue

**Comment by lucasjellema** [comment]
Dear @sijms ,

I am not quite sure what I did different from last time - but it now works! I have been able to interact with Autonomous Database using wallet and go_orav2 and nothing else.  So I am quite happy now.  (the trace file is now written as well).

I am working on an article in which I describe my experiences and share the code that works for me. I will share the URL for this article once done.  

Again, thanks for your help and guidance and for your kind offer to help investigate further using my wallet. It looks like I will not have to take you up on that offer, but I do sure appreciate it.

Here is my code that is now working successfully - and apparently with rather better performance than the implementation using the Oracle Instant Client.

```
package main

import (
	"context"
	"database/sql/driver"
	"fmt"
	"io"
	"time"

	go_ora "github.com/sijms/go-ora/v2"
)

func main() {
	service := "k8j2fvxbaujdcfy_daprdb_low.adb.oraclecloud.com"
	user := "demo"
	server := "adb.us-ashburn-1.oraclecloud.com"
	port := 1522
	pw := "Modem123mode"
	walletLocation := "/home/lucas/dapr-work/components-contrib/state/oracledatabase/Wallet_daprDB/"

	urlOptions := map[string]string{
		"TRACE FILE": "trace.log",
		"SSL":        "enable",
		"SSL Verify": "false",
		"WALLET":     walletLocation,
	}
	databaseURL := go_ora.BuildUrl(server, port, service, user, pw, urlOptions)
	conn, err := go_ora.NewConnection(databaseURL)
	if err != nil {
		fmt.Println("Can't create connection: ", err)
		return
	}
	err = conn.Open()
	if err != nil {
		fmt.Println("Can't open connection: ", err)
		return
	}
	defer func() {
		err = conn.Close()
		if err != nil {
			fmt.Println("Can't close connection: ", err)
		}
	}()
	var ctx = context.Background()
	err = conn.Ping(ctx)
	if err != nil {
		panic(fmt.Errorf("error pinging db: %w", err))
	}
	err = queryData(conn)
	if err != nil {
		panic(fmt.Errorf("error querying db: %w", err))
	}
	handleQuery("select to_char(systimestamp,'DD-MM-YYYY HH24:MI:SS') as the_time from dual", conn)
	handleQuery("select 42 as magic_number from dual", conn)

	executeStatement(createTableStatement, conn)
	executeStatement("INSERT INTO TEMP_TABLE ( NAME , VALUE) VALUES ('Johnny',391)", conn)
	executeStatement("INSERT INTO TEMP_TABLE ( NAME , VALUE) VALUES ('Marianne',312)", conn)
	handleQuery("select name, creation_time, value from TEMP_TABLE", conn)

	executeStatement(dropTableStatement, conn)
}


func Header(columns []string) {

}

func Record(columns []string, values []driver.Value) {
	for i, c := range values {
		fmt.Printf("%-25s: %v\n", columns[i], c)
	}
	fmt.Println()
}
func handleQuery(query string, conn *go_ora.Connection) {
	stmt := go_ora.NewStmt(query, conn)

	defer stmt.Close()

	rows, err := stmt.Query(nil)
	handleError("Can't query", err)
	defer rows.Close()

	columns := rows.Columns()

	values := make([]driver.Value, len(columns))

	Header(columns)
	for {
		err = rows.Next(values)
		if err != nil {
			break
		}
		Record(columns, values)
	}
	if err != io.EOF {
		handleError("Can't Next", err)
	}
}

func queryData(conn *go_ora.Connection) error {
	t := time.Now()

	stmt := go_ora.NewStmt("SELECT SYSTIMESTAMP FROM DUAL", conn)
	rows, err := stmt.Query_(nil)
	if err != nil {
		return err
	}
	defer func() {
		err = stmt.Close()
		if err != nil {
			fmt.Println("Can't close connection: ", err)
		}
	}()
	var Date time.Time
	for rows.Next_() {
		err = rows.Scan(&Date)
		if err != nil {
			return err
		}
		fmt.Println("Current Timestamp in Database: ", Date)
	}
	fmt.Println("Finish query rows: ", time.Now().Sub(t))
	return nil
}

func executeStatement(sqlText string, conn *go_ora.Connection) {
	stmt := go_ora.NewStmt(sqlText, conn)
	var err error
	defer func() {
		err = stmt.Close()
		if err != nil {
			fmt.Println("Can't close stmt", err)
		}
	}()

	result, err := stmt.Exec(nil)
	if err != nil {
		fmt.Println("Can't execute sql", err)
		return
	}
	rowsAffected, _ := result.RowsAffected()
	fmt.Println("Rows affected: ", rowsAffected)
}

func handleError(msg string, err error) {
	if err != nil {
		fmt.Println(msg, err)
		os.Exit(1)
	}
}

const createTableStatement = "CREATE TABLE TEMP_TABLE ( NAME VARCHAR2(100), CREATION_TIME TIMESTAMP DEFAULT SYSTIMESTAMP, VALUE  NUMBER(5))"
const dropTableStatement = "DROP TABLE TEMP_TABLE PURGE"
``` 

kind regards
Lucas

**Reply by sijms** [reply]
you are welcome

**Comment by sijms** [comment]
also note that PREFETCH_ROWS value dramatically affect performance you can set it as follow
if you are working on local network give it high value > 5000
if you are working on slow network keep it between 100 - 500
if you have any BLOB or CLOB column keep it default = 25

for now it is hardcoded to 25 until you specify a value in database URL
next commit I will try to calculate it automatically according to column size and presence of BLOB and CLOB to get better performance with each query

**Comment by lucasjellema** [comment]
Here is the link that I published on Medium regarding the connection to Automous Database: [link to article](https://lucasjellema.medium.com/connecting-go-application-to-oracle-database-on-prem-and-autonomous-with-and-without-oracle-c5535bbec135?sk=200861173a62acf1b631afeddb0a1441 )

I have also submitted a PR to the Dev branch in this repo with an example for connecting to a database using a wallet .

**Reply by sijms** [reply]
thank you for your article. it is really impressive

**Comment by sijms** [comment]
I read from your article that you facing  a problem related to reading "/" in the wallet path when passed in the database url
this is the buildURL code as you see I use QueryEscape for user, password, service and URL parameters
```golang
func BuildUrl(server string, port int, service, user, password string, options map[string]string) string {
	ret := fmt.Sprintf("oracle://%s:%s@%s:%d/%s", url.QueryEscape(user), url.QueryEscape(password),
		url.QueryEscape(server), port, url.QueryEscape(service))
	if options != nil {
		ret += "?"
		for key, val := range options {
			val = strings.TrimSpace(val)
			for _, temp := range strings.Split(val, ",") {
				temp = strings.TrimSpace(temp)
				if strings.ToUpper(key) == "SERVER" {
					ret += fmt.Sprintf("%s=%s&", key, temp)
				} else {
					ret += fmt.Sprintf("%s=%s&", key, url.QueryEscape(temp))
				}
			}
		}
		ret = strings.TrimRight(ret, "&")
	}
	return ret
}
```

**Comment by sijms** [comment]
I add new release that give some improvement with reading lob + automatically calculate PREFETCH_ROWS value
